logger.js

```diff
@@ -10,135 +10,142 @@
-var strftime = require('strftime').utc();
-
-require('colors');
-
-module.exports = function(config) {
-	config = config || {};
-	var exports = {};
+const { format } = winston;
 
-	config.levels = config.levels || {
-		none: 99,
-		trace: 0,
-		debug: 1,
-		log: 2,
+const customLevels = {
+	levels: {
+		fatal: 0,
+		error: 1,
+		warn: 2,
 		info: 3,
-		warn: 4,
-		error: 5,
-		fatal: 6,
-	};
-
-	config.level_abbr = config.level_abbr || {
-		trace: 'trc',
-		debug: 'dbg',
-		log: 'log',
-		info: 'inf',
-		warn: 'WRN',
-		error: 'ERR',
-		fatal: 'FTL',
-	};
-
-	config.filename = `${process.cwd()}/${config.filename || 'logs.log'}`;
-
-	config.errorLevel = config.errorLevel || 'log';
-
-	child_process.execSync(`mkdir -p ${path.dirname(config.filename)}`);
-	var log_file = fs.createWriteStream(config.filename, { flags: 'a' });
+		debug: 4,
+		query: 5,
+		trace: 10,
+		none: 99,
+	},
+	colors: {
+		fatal: 'yellow',
+		error: 'red',
+		warn: 'meganta',
+		info: 'blue',
+		debug: 'green',
+		query: 'green',
+		trace: 'cyan',
+		none: 'black',
+	},
+};
 
-	exports.setLevel = function(errorLevel) {
-		config.errorLevel = errorLevel;
-	};
+const logRootKeys = [
+	'level',
+	'timestamp',
+	'module',
+	'message',
+	'action',
+	'hostname',
+	'durationMs',
+];
+
+winston.addColors(customLevels.colors);
+
+const defaultLoggerConfig = {
+	filename: 'logs/lisk.log',
+	level: 'none',
+	consoleLevel: 'debug',
+};
 
-	function snipFragileData(data) {
-		for (var key in data) {
-			if (key.search(/passphrase|password/i) > -1) {
-				data[key] = 'XXXXXXXXXX';
+const formatInfo = format(info =>
+	Object.keys(info).reduce((accumulated, key) => {
+		if (logRootKeys.includes(key)) {
+			accumulated[key] = info[key];
+		} else {
+			if (!accumulated.meta) {
+				accumulated.meta = {};
 			}
+			accumulated.meta[key] = info[key];
 		}
-		return data;
-	}
-
-	Object.keys(config.levels).forEach(name => {
-		function log(message, data) {
-			var log = {
-				level: name,
-				timestamp: strftime('%F %T', new Date()),
-			};
+		return accumulated;
+	}, {}));
+
+const consoleFormat = info => {
+	let log = `[${info.level}] ${info.timestamp} <${info.module}>: ${
+		info.message
+	}`;
+	log += info.durationMs ? ` duration=${info.durationMs}ms` : '';
+	log += info.meta ? `\n\t ${JSON.stringify(info.meta)}` : '';
+	return log;
+};
 
-			if (message instanceof Error) {
-				log.message = message.stack;
-			} else {
-				log.message = message;
-			}
+const injectKey = (key, value) =>
+	format(info => {
+		info[key] = value;
+		return info;
+	});
 
-			if (data && util.isObject(data)) {
-				log.data = JSON.stringify(snipFragileData(data));
-			} else {
-				log.data = data;
-			}
+const consoleLog = (level, module) =>
+	new winston.transports.Console({
+		level,
+		levels: customLevels.levels,
+		format: format.combine(
+			format.colorize(),
+			injectKey('module', module)(),
+			format.timestamp(),
+			formatInfo(),
+			format.printf(consoleFormat)
+		),
+	});
```

Comment:
I really like bunyan CLI approach here. It just take JSON object from `STDOUT` and format it outside the main process. This avoid performance overhead for just formatting logs to show in console.

Reply:
>  This avoid performance overhead for just formatting logs to show in console.
console log shouldnt be turned on the server or at least with less level.

Reply:
Still its not the right way to do it. You can use `tail -f lisk.log | bunyan` and it will show you logs on screen in the way you normally see in console. So it does not matter if you enabled console or not, formatting logs for console outside main process is helpful.